### PR TITLE
octave: bump revision (gcc)

### PR DESCRIPTION
**Pull request by siko1056**
- [x] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/HEAD/CONTRIBUTING.md)?
- [x] Have you ensured that your commits follow the [commit style guide](https://docs.brew.sh/Formula-Cookbook#commit)?
- [x] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [ ] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [ ] Is your test running fine `brew test <formula>`, where `<formula>` is the name of the formula you're submitting?
- [ ] Does your build pass `brew audit --strict <formula>` (after doing `brew install --build-from-source <formula>`)? If this is a new formula, does it pass `brew audit --new <formula>`?

-----

The gcc of the "octave" bottle is outdated. gcc version 11.1.0 is hard coded inside, while an updated system uses gcc version 11.2.0 already.  This issue is fixed by rebuilding the bottle with bumping the revision.

The stale gcc link causes very frequently used Octave packages to not install correctly.  Entering the first line in the Octave command-line results in the following error messages, clearly indicating the problem of a stale gcc compiler.

```
>> pkg install -forge control

ld: warning: directory not found for option '-L/usr/local/opt/gcc/lib/gcc/11/gcc/x86_64-apple-darwin20/11.1.0'
ld: warning: directory not found for option '-L/usr/local/opt/gcc/lib/gcc/11/gcc/x86_64-apple-darwin20/11.1.0/../../..'
ld: library not found for -lgfortran
clang: error: linker command failed with exit code 1 (use -v to see invocation)
make: *** [__control_slicot_functions__.oct] Error 1
tar -xzf slicot.tar.gz
/usr/local/Cellar/octave/6.3.0/bin/mkoctfile-6.3.0 -Wall -Wno-deprecated-declarations  __control_helper_functions__.cc
mkdir sltmp
mv slicot/src/*.f ./sltmp
mv slicot/src_aux/*.f ./sltmp
if [ "1" = "1" ]; then \
                echo "copy routines using DGGES"; \
                cp SB04OD.fortran ./sltmp/SB04OD.f; \
                cp SG03AD.fortran ./sltmp/SG03AD.f; \
                cp SG03BD.fortran ./sltmp/SG03BD.f; \
        fi;
copy routines using DGGES
cp AB08NX.fortran ./sltmp/AB08NX.f
cp AG08BY.fortran ./sltmp/AG08BY.f
cp SB01BY.fortran ./sltmp/SB01BY.f
cp SB01FY.fortran ./sltmp/SB01FY.f
cp SB06ND.fortran ./sltmp/SB06ND.f
cp TB01MD.fortran ./sltmp/TB01MD.f
cp TB01ND.fortran ./sltmp/TB01ND.f
cp TB01ZD.fortran ./sltmp/TB01ZD.f
cp TG04BX.fortran ./sltmp/TG04BX.f
cp ODLTZM.fortran ./sltmp/ODLTZM.f
cp makefile.slicot ./sltmp/makefile
cd sltmp; /usr/local/Cellar/octave/6.3.0/bin/mkoctfile-6.3.0 -w -c MA02ID.f; rm MA02ID.f; /usr/local/Cellar/octave/6.3.0/bin/mkoctfile-6.3.0 -c *.f;
ar -rc slicotlibrary.a ./sltmp/*.o
rm -rf sltmp slicot
LDFLAGS="-F/usr/local/opt/qt@5/lib  -L/usr/local/opt/openblas/lib -lopenblas  -L/usr/local/opt/gcc/lib/gcc/11/gcc/x86_64-apple-darwin20/11.1.0 -L/usr/local/opt/gcc/lib/gcc/11/gcc/x86_64-apple-darwin20/11.1.0/../../.. -lgfortran -lquadmath -lm" \
    /usr/local/Cellar/octave/6.3.0/bin/mkoctfile-6.3.0 -Wall -Wno-deprecated-declarations__control_slicot_functions__.cc common.cc slicotlibrary.a

error: pkg: error running 'make' for the control package.
error: called from
    configure_make at line 110 column 9
    install at line 196 column 7
    pkg at line 568 column 9
```

A better approach would probably be to automatically rebuild this bottle if gcc is updated.  However, this seems to be not desired and I do not understand the following logic, which tried to solve this issue:

https://github.com/Homebrew/homebrew-core/blob/250d6e6063d18341fe81255e97f9647554e98937/Formula/octave.rb#L125-L129

Detailed discussion: https://github.com/Homebrew/discussions/discussions/1986

### CODE REVIEW

**Comment by cho-m** [comment]
Probably should see if there is a way for current `-L/usr/local/opt/gcc/lib/gcc/11/gcc/x86_64-apple-darwin20/11.1.0/../../..` to instead traverse the `..`s and use `-L/usr/local/opt/gcc/lib/gcc/11`, where `libgfortan` is available.

Otherwise, at minimum, there needs to be a test case added to detect this. When `gcc` is updated, dependents like `octave` will be tested, so the only way to know when to do a revision bump is if the test fails.

**Comment by SMillerDev** [comment]
>  However, this seems to be not desired and I do not understand the following logic, which tried to solve this issue

This should replace the versioned path `/usr/local/Cellar/11.1.0/gcc/` with the unversioned `/usr/local/opt/gcc/`, which should solve that issue. Unfortunately it seems that there is a version a little further in the include path too.

Like mentioned in the discussion, could you add a test that fails if GCC gets updated to a version that octave doesn't have baked in?

**Comment by siko1056** [comment]
Thank you @cho-m and @SMillerDev  for the comments.  I will work on a small example and extend this PR.

**Comment by carlocab** [comment]
One alternative here is to force GCC to stop encoding its full version in its subdirectories. I vaguely recall a `configure`-time option that enables this.

This will break any existing users that rely on those paths, but they get broken with each version bump anyway.

**Comment by carlocab** [comment]
#84248

**Comment by SMillerDev** [comment]
Regardless of the fix, it would still be good to have a check for it.

**Comment by siko1056** [comment]
Thank you @carlocab for the quick help and input :slightly_smiling_face: 

With  55cad12 I added two minimal compilations tests.  The second is the most minimal version of the problem I could think of.

Please tell me if my "rusty" ruby syntax knowledge for long strings is adequate and I will polish the PR if necessary :wink:

File: Formula/octave.rb [hunk 1]
```diff
@@ -141,11 +141,9 @@ def install
...
     system bin/"octave", "--eval", "single ([1+i 2+i 3+i]) * single ([ 4+i ; 5+i ; 6+i])"
     # Test basic compilation
-    system cat<<EOT >> oct_demo.cc
-    #include <octave/oct.h>
-    DEFUN_DLD (oct_demo, args, /*nargout*/, "doc str")
-    { return ovl (42); }
-    EOT
+    (testpath/"oct_demo.cc").write("#include <octave/oct.h>\n"\
+    "DEFUN_DLD (oct_demo, args, /*nargout*/, \"doc str\")"\
+    "{ return ovl (42); }\n")
```
Comment (carlocab) [inline Formula/octave.rb]:
```suggestion
    (testpath/"oct_demo.cc").write <<~EOS 
      #include <octave/oct.h>
      DEFUN_DLD (oct_demo, args, /*nargout*/, "doc str")
      { return ovl (42); }
    EOS
```

File: Formula/octave.rb [hunk 2]
```diff
@@ -139,5 +140,15 @@ def install
     system bin/"octave", "--eval", "(22/7 - pi)/pi"
     # This is supposed to crash octave if there is a problem with BLAS
     system bin/"octave", "--eval", "single ([1+i 2+i 3+i]) * single ([ 4+i ; 5+i ; 6+i])"
+    # Test basic compilation
+    (testpath/"oct_demo.cc").write("#include <octave/oct.h>\n"\
+                                   "DEFUN_DLD (oct_demo, args, /*nargout*/, \"doc str\")"\
+                                   "{ return ovl (42); }\n")
```
Comment (carlocab) [inline Formula/octave.rb:146]:
```suggestion
    (testpath/"oct_demo.cc").write <<~EOS
      #include <octave/oct.h>
      DEFUN_DLD (oct_demo, args, /*nargout*/, "doc str")
      { return ovl (42); }
    EOS
```

File: Formula/octave.rb [hunk 3]
```diff
@@ -139,5 +140,15 @@ def install
...
+    system bin/"octave", "--eval", "mkoctfile('oct_demo.cc'); assert(oct_demo, 42)"
+    # Test FLIBS environment variable
+    system bin/"octave", "--eval", "args = strsplit (mkoctfile ('-p', 'FLIBS')); "\
+                                   "args = args(~cellfun('isempty', args));"\
+                                   "mkoctfile (args{:}, 'oct_demo.cc'); "\
+                                   "assert(oct_demo, 42)"
```
Comment (carlocab) [inline Formula/octave.rb:152]:
```suggestion
    system bin/"octave", "--eval", <<~EOS
      args = strsplit (mkoctfile ('-p', 'FLIBS'));
      args = args(~cellfun('isempty', args));"
      mkoctfile (args{:}, 'oct_demo.cc'); "
      assert(oct_demo, 42)
    EOS
```

**Comment by siko1056** [comment]
Thanks for the suggestions @carlocab , your ruby syntax looks indeed cleaner :wink:

File: Formula/octave.rb [hunk 4]
```diff
@@ -141,14 +141,18 @@ def install
...
     system bin/"octave", "--eval", "mkoctfile('oct_demo.cc'); assert(oct_demo, 42)"
     # Test FLIBS environment variable
-    system bin/"octave", "--eval", "args = strsplit (mkoctfile ('-p', 'FLIBS')); "\
-                                   "args = args(~cellfun('isempty', args));"\
-                                   "mkoctfile (args{:}, 'oct_demo.cc'); "\
-                                   "assert(oct_demo, 42)"
+    system bin/"octave", "--eval", <<~EOS
+      args = strsplit (mkoctfile ('-p', 'FLIBS'));
+      args = args(~cellfun('isempty', args));"
+      mkoctfile (args{:}, 'oct_demo.cc'); "
```
Comment (carlocab) [inline Formula/octave.rb:154]:
```suggestion
      args = args(~cellfun('isempty', args));
      mkoctfile (args{:}, 'oct_demo.cc');
```
Are the `"`s still needed?

Comment (siko1056) [reply Formula/octave.rb:154]: It is not needed anymore.

**Comment by siko1056** [comment]
@carlocab Are further changes necessary here?  The test failure seems to be related to a general compilation problem of Octave on macOS 11 arm64, the remaining tests were cancelled :sweat:

**Comment by carlocab** [comment]
> I think we want to wait for #84248 to run this test again

Actually, I think it's the reverse. We want to merge this first so that we can check that the test here fails after the revision bump to GCC in #84248.

If the test doesn't catch the change in paths from the GCC change then it's not the test we need.

**Comment by siko1056** [comment]
@SMillerDev @carlocab is there anything I can help on this item?

**Comment by carlocab** [comment]
Fix the failing test, as merging this as is will lead to all the bottles going away.

**Comment by iMichka** [comment]
@siko1056 I think the test might still be of interest. It depends on how much work it is do make it run. Do you need any help to get a green CI here?

**Comment by siko1056** [comment]
Thanks for your reply @iMichka . The last test run is quite some time ago.  If it was run again, I think it should pass now. Can you trigger the test suit run again?

The more interesting question is, does it fail if GCC links are stale again, i.e. the purpose of the test 😉

**Comment by siko1056** [comment]
@iMichka I made some updates and only the first time the test suite was run.  Can you trigger another run?  Many thanks in advance :slightly_smiling_face:

**Comment by iMichka** [comment]
You have a syntax error:
```
Formula/octave.rb:149:1: C: [Correctable] Use 2 spaces for indentation in a heredoc.
        mkoctfile('-v', '-std=c++11', 'oct_demo.cc'); ...
```

**Comment by siko1056** [comment]
Thanks @iMichka, I fixed the problem and some tests pass now.  However, some tests timed out and on "ubuntu-latest" the "mkoctfile" configuration seems not to be setup correctly.

If the workers are less busy, can the check be re-run? :sweat:

**Comment by siko1056** [comment]
Thank you @carlocab for the extra attempt.  All checks except unbuntu are working, enough to merge? :innocent:

**Comment by carlocab** [comment]
Not quite:
1. Merging PRs with `brew test` failures breaks bottles.
2. You still have a merge commit on your branch, so BrewTestBot won't be able to merge this (Homebrew/core has a linear history requirement).

We can probably fix the first one by adding appropriate `-L` flags. Something like `-L#{lib}/octave/#{version}` should work. Failing that, we can skip the last part of the test on Linux with `return if OS.linux?`, but we should really try to fix it first.

You can fix the second one by rebasing your branch against origin/master. Please squash all your commits too when you do so.

**Comment by siko1056** [comment]
Thanks for the constructive comments @carlocab .

- Fixed item 2: I rebased my patch-1-branch and sqashed everything into a single commit.

- Item 1: I don't know, why `mkoctfile` "forgets" all during the installation set "-L" flags?  So far I only use Homebrew for macOS, but I try to use it for my Linux and figure out what is going on...

**Comment by iMichka** [comment]
> octave: no graphical display found

This is because we run the tests in a docker container on Linux.

Can we try to run all the `octave` calls in the test block with the `-W` flag, which should disable the GUI part of `octave`?

**Comment by siko1056** [comment]
Thanks for checking @iMichka .  In general Octave does not really care if there is no display available and just turns of GUI related features as the output suggests
```
octave: no graphical display found
octave: disabling GUI features
```
The problem for the Linux compilation check is this error:
```
/home/linuxbrew/.linuxbrew/bin/ld: cannot find -loctinterp
/home/linuxbrew/.linuxbrew/bin/ld: cannot find -loctave
```
resulting from
https://github.com/Homebrew/homebrew-core/runs/4345829738?check_suite_focus=true#step:7:55
```
/home/linuxbrew/.linuxbrew/bin/g++-11
  -I/home/linuxbrew/.linuxbrew/Cellar/octave/6.4.0_1/include/octave-6.4.0/octave/..
  -I/home/linuxbrew/.linuxbrew/Cellar/octave/6.4.0_1/include/octave-6.4.0/octave
  -I/home/linuxbrew/.linuxbrew/Cellar/octave/6.4.0_1/include
  -pthread -fopenmp -Os -w -pipe -march=native  -std=c++11
  -o oct_demo.oct  /tmp/oct-pF7pis.o
  -shared -Wl,-Bsymbolic
  -L/home/linuxbrew/.linuxbrew/lib
  -L/home/linuxbrew/.linuxbrew/Cellar/octave/6.4.0_1/lib
  -loctinterp -loctave
```
I try to find out the contents of those `-L` directories on the builder machine.  They should contain `liboctinterp.so` and `liboctave.so` to sucessfully compile the test file.

**Comment by carlocab** [comment]
`liboctave.so` and `liboctinterp.so` are in `/home/linuxbrew/.linuxbrew/Cellar/octave/6.4.0_1/lib/octave/6.4.0`, which is why I suggested you add the flag `-L#{lib}/octave/#{version}`

**Comment by siko1056** [comment]
@carlocab Thank you again for this enlightning comment :slightly_smiling_face:  Homebrew's build system is much better than I expected, great work you do :clap: 

Now I (hopefully) understand your previous comment regarding the `-L#{lib}/octave/#{version}` and lets wait for the builders to finish :coffee:

**Comment by siko1056** [comment]
It seems to have worked :tada:  Thank you for all your helpful comments.

Ready to merge? :innocent:

**Review by carlocab — APPROVED** [review]
Thanks for working on this!

**Review by iMichka — APPROVED** [review]

**Comment by iMichka** [comment]
Nice, thanks!

**Pull request closed without merging**